src/databricks/labs/ucx/azure/access.py

```diff
@@ -54,28 +57,75 @@ def __init__(
             "Storage Blob Data Owner": Privilege.WRITE_FILES,
             "Storage Blob Data Reader": Privilege.READ_FILES,
         }
+        self._permission_levels = {
+            "Microsoft.Storage/storageAccounts/blobServices/containers/write": Privilege.WRITE_FILES,
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/write": Privilege.WRITE_FILES,
+            "Microsoft.Storage/storageAccounts/blobServices/containers/read": Privilege.READ_FILES,
+            "Microsoft.Storage/storageAccounts/blobServices/containers/blobs/read": Privilege.READ_FILES,
+        }
+
+    def _get_permission_level(self, permission_to_match: str) -> Privilege | None:
+        # String might contain '*', check for wildcard match
+        pattern = re.sub(r'\*', '.*', permission_to_match)
+        permission_compiled = re.compile(pattern)
+        for each_level, privilege_level in self._permission_levels.items():
+            # Check for storage blob permission with regex to account for star pattern
+            match = permission_compiled.match(each_level)
+            if match:
+                # If a write permission is found, no need to check for read permissions
+                return privilege_level
+        return None
+
+    def _get_custom_role_privilege(self, role_permissions: list[str]) -> Privilege | None:
+        # If both read and write privileges are found, only write privilege will be considered
+        higher_privilege = None
+        for each_permission in role_permissions:
+            privilege = self._get_permission_level(each_permission)
+            if not privilege:
+                continue
+            # WRITE_FILES is the higher permission, don't need to check further
+            if privilege == Privilege.WRITE_FILES:
+                return privilege
+            if privilege == Privilege.READ_FILES:
+                higher_privilege = privilege
+        return higher_privilege
+
+    def _get_role_privilege(self, role_assignment: AzureRoleAssignment) -> Privilege | None:
+        privilege = None
+        # Check for custom role permissions on the storage accounts
+        if role_assignment.role_permissions:
+            privilege = self._get_custom_role_privilege(role_assignment.role_permissions)
+        elif role_assignment.role_name in self._levels:
+            privilege = self._levels[role_assignment.role_name]
```

Comment:
this mean custom roles take higher priority than pre-defined role, which could be wrong. the cleanest approach is to collate all the output into a set and then emit WRITE_FILES if it is in the set, else emit READ_FILES.

Reply:
@nkvuong Thanks for reviewing. One role assignment can either be Custom, or Built-in. If one SP (sp1) has 2 role assignments (one custom read and one built-in write) for a container, then the first iteration in the loop will put READ in the principal_spm_mapping dict. The next iteration will find the write access from built-in and update the dict with the WRITE access.
I have added this scenario to unit test test_save_spn_permissions_custom_role_valid_azure_storage_account() for Custom_role_11. Please review and let me know if this doesn't seem correct.

Reply:
@nkvuong Please let me know if this makes sense, or if some changes should be made.